src/Components/Web.JS/src/Rendering/BrowserRenderer.ts

```diff
@@ -263,6 +262,8 @@ export class BrowserRenderer {
       const selectValue: string | null = newDomElementRaw[selectValuePropname];
       setSelectElementValue(newDomElementRaw, selectValue);
     }
+
+    insertLogicalChild(newDomElementRaw, parent, childIndex);
```

Comment:
I think it might not be enough to do this only here. Note that on line 251, it recurses into the entire rest of the DOM building process. So the change as-is here would invert the ordering of building the entire DOM, which might have large numbers of unintended consequences. It *might* work fine, or it might cause big problems depending on whether (for example) custom elements care about being attached to the document at the time they are inserted into their parent.

To avoid this problem, I recommend calling `insertLogicalChild` in two places instead of just one:

 * Immediately before the call to `this.insertFrameRange` on line 251. This will apply to elements with descendants.
 * Immediately after the `for` loop completes on line 254. This will apply to elements with no descendants.

Of course, we only want to make one of these two calls on any given pass, so you'll also need a flag to track whether it was called on line 251 so you can then skip it on line 254.

Reply:
Also it's worth clarifying: when a custom element is attached in regular HTML documents, are its descendants all populated by that time, or is it just its attributes?

If you expect to see all the *descendants* already in the document, the logic you have already would be correct. Whereas if you expect to see all the *ancestors* already in the document, then the logic you have here is not correct and should be updated as in my comment above.

Reply:
You're right, I missed the recursive part. I'd say the user will expect ancestors to be ready.

Reply:
On second thought: Every DOM insert is a potential cause of reflow. Leaf to trunk insertion might be benificial and way cheaper than the current order.

Reply:
I don't think it will reflow during the synchronous event loop, but leaf-to-trunk might be beneficial if it looks more like normal HTML from the point of view of the custom element.

It depends on when the browser calls `connectedCallback`. Does it do it [A] when the element is inserted into its parent, or [B] when its ancestor is attached to the real live document?

If it's [A], I think we need trunk-to-leaf order so that we don't trigger these notifications before inserting into the real document.

If it's [B], then I think leaf-to-trunk is better because then the custom element gets to see all its ancestors and descendants at once during `connectedCallback`.

Reply:
I tested and it's synchronous. So no reflows during JS execution. Also connectedCallback is executed when the element is attachted to an element that is part of the document.

Reply:
Actually I've just realised that `tryApplyValueProperty` relies on being able to scan up the DOM hierarchy to find the closest `<select>` ancestor. So we can't call insert leaf-to-trunk without breaking this. In fact we can't even apply attributes before inserting the element without breaking this.

Reply:
> 
> 
> Actually I've just realised that `tryApplyValueProperty` relies on being able to scan up the DOM hierarchy to find the closest `<select>` ancestor. So we can't call insert leaf-to-trunk without breaking this. In fact we can't even apply attributes before inserting the element without breaking this.

Why not setting the "selected" attribute on the OPTION. This seems to be the official way to select an option of a select?

Reply:
Good question :) It's because there are different insertion orders to account for. Hopefully this will clarify: https://github.com/dotnet/aspnetcore/blob/master/src/Components/Web.JS/src/Rendering/BrowserRenderer.ts#L257-L261

Reply:
> Good question :) It's because there are different insertion orders to account for. Hopefully this will clarify: https://github.com/dotnet/aspnetcore/blob/master/src/Components/Web.JS/src/Rendering/BrowserRenderer.ts#L257-L261

I got that ;-)

But the question is: Do we want to support HTMLSelectElement.value from Blazor? It's a string that represents the value that is going to be sent on form submission. normally a "preselected" value (or multiple) would be indicated by a selected attribute on the option. The fact that you can get / set it from javascript does not mean that it needs to be supported at DOM insertion. It's not official behaviour anyway.

This would be my way to handle selecting an option value (possibly delayed):

```
<form>
<select>
    @foreach (var item in items)
    {
        <option value="@item.Item1" selected="@item.Item2">@item.Item1</option>
    }
</select></form>

@code
{
    List<(string, bool)> items = new List<(string, bool)>()
    {
        ("Test", false),
        ("Test2", false),
    };

    protected override async Task OnInitializedAsync()
    {
        await Task.Delay(3000);
        items.Add(("Test3", true));
        await this.InvokeAsync(() => this.StateHasChanged());
    }
}
```

Reply:
We need to support all existing syntaxes, including binding:

```
<select @bind="myValue">
    <option>First</option>
    <option>Second</option>
    <option>Third</option>
</select>
```

Reply:
What I suspect we'd need to do, in order to switch to the leaf-to-trunk order, is to go through all of the logic involved in building the DOM (unfortunately!) to find exactly which parts of it might rely on the previous trunk-to-leaf ordering. It might be that running the E2E tests highlights most of those cases, but it would be worth attempting some manual inspection too.

For any such case, we would need to come up with a way of flowing in whatever information is needed to make it work. For example, when applying values to a `<select>`, we'd need the `<option>` elements we're inserting to be able to trace back up the DOM to find their closest `<select>`, even if they haven't yet been inserted into the document. It might be possible to do this by changing something like `LogicalElement` to have a concept of "what my parent node is or will be once I'm inserted into the physical DOM" so that the logic for tracing up the hierarchy can still work even when the chain of parents isn't yet physically established.

Overall I recognize this is pretty complex stuff. If you think it's impractical to attempt I completely understand. However if you are keen to have a go I'm happy to try to help you along!

Reply:
I went at it while preserving the original order. Reversing the order is for another time.